scripts: sanitycheck: Generate json report format

yerabolu opened the pull request:
Generating JSON report format without changing
the existing report formats.
JSON report is divided into three sections
environment tests are run
test cases basic info and discarded tests
after applying filters
Added the complete debug log to the report

Fixes: #26443

Signed-off-by: Spoorthy Priya Yerabolu <spoorthy.priya.yerabolu@intel.com>

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3229,6 +3232,45 @@ def csv_report(self, filename):
+                rowdict["testspec"].append({"ram_size": instance.metrics.get("ram_size", 0),
+                                            "rom_size": instance.metrics.get("rom_size",0)
+                                            })
+        rowdict["apply_filters"] = {"discards": str(self.apply_filters())}
```
this is just dumping a dictionary into json, the data from discards should be per instance in this case, looks like this in json which is not ok:

```
{
   "apply_filters" : {
      "discards" : "{<TestCase samples/hello_world/sample.basic.helloworld on esp32>: 'Not supported by the toolchain', <TestCase samples/hello_world/sample.basic.helloworld on xt-sim>: 'Not supported by the toolchain', <TestCase samples/hello_world/sample.basic.helloworld on esp32>: 'Not supported by the toolchain', <TestCase samples/hello_world/sample.basic.helloworld on qemu_x86_nommu>: 'Excluded tags per platform (only_tags)', <TestCase samples/hello_world/sample.basic.helloworld on qemu_x86_coverage>: 'Excluded tags per platform (only_tags)', <TestCase samples/hello_world/sample.basic.helloworld on qemu_x86_xip>: 'Excluded tags per platform (only_tags)', <TestCase samples/hello_world/sample.basic.helloworld on nrf52_bsim>: 'Environment (BSIM_OUT_PATH) not satisfied', <TestCase samples/hello_world/sample.basic.helloworld on stm32f030_demo>: 'Not enough RAM',
```

yerabolu replied:
addressed it, now it adds per instance and details.

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3229,6 +3232,45 @@ def csv_report(self, filename):
+                                            "rom_size": instance.metrics.get("rom_size",0)
+                                            })
+        rowdict["apply_filters"] = {"discards": str(self.apply_filters())}
+        with open(os.path.join(ZEPHYR_BASE, "sanity-out", "sanitycheck.log"), "r") as log_file:
```
This is attaching the sanitycheck log which basically had everything in verbose mode from running the command, not sure how this is useful, we have the file already, the logs should be per instance(build.log, handler.log and device.log when something went wrong, I would have handler.log all the time, even in case of success.

yerabolu replied:
per instance it adds the handler and build logs

yerabolu replied:
missed adding device logs, working on adding it.

yerabolu replied:
Added device logs when the tests fails or errors out.

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3229,6 +3232,45 @@ def csv_report(self, filename):
+
+    def json_report(self, filename):
+        rowdict = {}
+        rowdict["environment"] = {"OS": os.name,
```
captialized here, later all keys are small cap.

yerabolu replied:
changed it to be consistent

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3229,6 +3232,45 @@ def csv_report(self, filename):
+                                        })
+            if instance.status not in ["failed", "timeout"]:
+                if instance.handler:
+                    rowdict["testspec"].append({"handler_time": (instance.metrics.get           ("handler_time", 0))})
```
those are not associated with any instance.

nashif commented:
as test for the output, you should be able to write a completely independent script that loads and parses the json file and then generate the same junit files generated right now by sanitycheck.

yerabolu commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3229,6 +3232,45 @@ def csv_report(self, filename):
                 cw.writerow(rowdict)
 
+    def set_default(self, obj):
+        ''' JSON doesn't recognize set type so need to converted to list '''
```
removed the default method and serialized it in the code.

yerabolu committed:
scripts: sanitycheck: Generate json report format

Generating JSON report format without changing
the existing report formats.
JSON report is divided into three sections
environment tests are run
test cases basic info and discarded tests
after applying filters
Added the complete debug log to the report

Fixes: #26443

Signed-off-by: Spoorthy Priya Yerabolu <spoorthy.priya.yerabolu@intel.com>

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3297,6 +3299,57 @@ def csv_report(self, filename):
+        rowdict["environment"] = {"os": os.name,
+                                    "toolchain": self.get_toolchain()
+                                    }
+        rowdict["discards"] = []
```
data from discard is now part of the overall report per test, so no need to have this as part of the output. when a test is skipped, there has to be a reason which is part of the instance.

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3297,6 +3299,57 @@ def csv_report(self, filename):
+                                            })
+            if instance.status in ["passed"]:
+                if instance.handler:
+                    rowdict["testspec"].append({"handler_time": (instance.metrics.get           ("handler_time", 0))})
```
you are now adding handler_time as a new list entry of the testspec which is not what we want, this should be in the same dict you started above.

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3297,6 +3299,57 @@ def csv_report(self, filename):
+            if instance.status in ["passed"]:
+                if instance.handler:
+                    rowdict["testspec"].append({"handler_time": (instance.metrics.get           ("handler_time", 0))})
+                    rowdict["testspec"].append({"ram_size": instance.metrics.get ("ram_size", 0),
```
same here

yerabolu replied:
Resolved it

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3297,6 +3299,57 @@ def csv_report(self, filename):
+                handler_log = os.path.join(log_path, "handler.log")
+                if os.path.exists(handler_log):
+                    rowdict["testspec"].append({"handler_log":self.process_log(handler_log)})
+
```
same here, you are adding all this data as new elements to the testspec list, that not what we want. Each element of the list shall be one single test.

nashif commented:
looking good, however, when a test is skipped, you do not capture the reason.

yerabolu commented:
> looking good, however, when a test is skipped, you do not capture the reason.

Captured skip reason

nashif reviewed (commented):
some feedback to the syntax now that we finally have the correct structure..

- you are using internal sanitycheck terminology here like harness and handler which should be changed to something that is more test result specific and easy to understand from a testing perspective. "Handler log" is not really clear in a test context, maybe should be something like test_output.

- Please be consistent with keys, I would prefer all lower case with no spaces.

- There is some data you are capturing that I am not sure provide any value from a test reporting perspective, while there is no harm including them, this will be just noise and information we could get from somewhere else, ie.. the testcase.yaml itself, examples are the name of the harness or the extra configs being used...

- Need to make this easily parseable, so if we do not have a value, I would just leave it empty, for example it is much easier to check if "Build log" key exist than checking with:

```
if d['Device log'] == "No device logs":
    ...
```

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3297,6 +3299,62 @@ def csv_report(self, filename):
+            elif instance.status in ["error", "failed", "timeout"]:
+                if os.path.exists(handler_log):
+                    json_dict["Handler log"] = self.process_log(handler_log)
+                    json_dict["Device log"] = self.process_log(device_log)
```
we do not have device_log for everything, so this could be empty

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3297,6 +3299,62 @@ def csv_report(self, filename):
+
+            if instance.status in ["passed"]:
+                if instance.handler:
+                    json_dict["Handler time"] =  instance.metrics.get("handler_time", 0)
```
need some better naming here, duration? execution_time?

yerabolu replied:
Changed it to execution_time.

PerMac commented:
The current version of the json output does not provide a tree structure we would like to have. Please compare for e.g. running `./scripts/sanitycheck -T tests/kernel/common/ -p native_posix./scripts/sanitycheck -T tests/kernel/common/ -p native_posix` and compare json report with native_posix.xml report. The XML one has a structure where you have test suites and inside there are all individual test cases. 
Please add the extra layer inside test suites with test cases.

PerMac reviewed (changes requested):
as commented

PerMac commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -2587,6 +2589,7 @@ def save_reports(self, name, suffix, report_dir, no_update, release, only_failed
             self.xunit_report(filename + ".xml", full_report=False, append=only_failed)
             self.xunit_report(filename + "_report.xml", full_report=True, append=only_failed)
             self.csv_report(filename + ".csv")
+            self.json_report(filename + ".json")
```
I see no handling of "only_failed" running mode. It allows rerunning only failed tests and modify the results of failed tests instead of rewriting the whole report. In the current version of PR if e.g. there were 8 test suites and one failed and one called sanitycheck with "--only-failed" the report with 8 suites will be replaced with the newest one with only one suite. self.xunit_report() deals with it by "append=only_failed".

nashif replied:
> I think there is a bug now in reporting @nashif. sanitycheck.xml report behaves the same as this json report. The summary entry states the correct number of tests executed but inside there are only entries from retried tests. Other reports (sanitycheck_report.xml and platform.xml) are correct, they have all the entries

file a bug please.

PerMac commented:
> as test for the output, you should be able to write a completely independent script that loads and parses the json file and then generate the same junit files generated right now by sanitycheck.

Let me expand on this commenting on what is missing IMO.
- As a tester (and maintainer of CIs) I would like to have easily available (i.e. saved as a key-value pairs) info which is standard to test reports summarizing the results. This is an example of an entry from junit:
```
<testsuite errors="262" failures="0" name="nrf52840dk_nrf52840" skipped="0" tests="750" time="244.814200">
```
I would like to have dictionary entries for this summary info. Currently, you have time, platform, but the information with the number of failures/errors/skipped/total is missing

This should be available on all levels: e.g one can have a "master" suite, which consists of 100 test suites ("instances" in sanitycheck's code nomenclature). There should be a summary with all the above information e.g {errors:1, failures:2, skipped:5, tests:100}". Then each test suite (instance) will have its own summary inside, e.g. test suite #3, which has 500 test cases, will have "{errors:0, failures:1, skipped:0, tests:500}".

PerMac commented:
The current standard for test verdicts is failure/error/pass/skipped. I would suggest using the same for now. I saw that you use a status key, but e.g. I saw that test which ended with timeout are reported in json as `timeout` and in junit as `failure` with details in the message informing that it was "timeout". I would prefer to have consistency between both reports. It can be achieved by using `instance.results` and translating them as `xunit_report()` method is doing. 
@nashif In the meant time I think we should consider bringing back the discussion about verdicts to use (raise during next testers wg meeting? or are we fine with what was agreed on?)

PerMac commented:
Tests can be very verbose. I would consider including handler.log only in case of failure

PerMac commented:
@nashif We should consider having an additional "target" json report as with platform.xml for ease of publishing results. But this can be added later on and in principle should be a one-liner change.

yerabolu commented:
> some feedback to the syntax now that we finally have the correct structure..
> 
> * you are using internal sanitycheck terminology here like harness and handler which should be changed to something that is more test result specific and easy to understand from a testing perspective. "Handler log" is not really clear in a test context, maybe should be something like test_output.
> * Please be consistent with keys, I would prefer all lower case with no spaces.
> * There is some data you are capturing that I am not sure provide any value from a test reporting perspective, while there is no harm including them, this will be just noise and information we could get from somewhere else, ie.. the testcase.yaml itself, examples are the name of the harness or the extra configs being used...
> * Need to make this easily parseable, so if we do not have a value, I would just leave it empty, for example it is much easier to check if "Build log" key exist than checking with:
> 
> ```
> if d['Device log'] == "No device logs":
>     ...
> ```

Submitted changes to the json format as suggested above. Than you for the detailed review.

yerabolu commented:
> The current version of the json output does not provide a tree structure we would like to have. Please compare for e.g. running `./scripts/sanitycheck -T tests/kernel/common/ -p native_posix` and compare json report with native_posix.xml report. The XML one has a structure where you have test suites and inside there are all individual test cases.
> Please add the extra layer inside test suites with test cases.

Add the extra layer with platform name.

yerabolu commented:
> The current version of the json output does not provide a tree structure we would like to have. Please compare for e.g. running `./scripts/sanitycheck -T tests/kernel/common/ -p native_posix` and compare json report with native_posix.xml report. The XML one has a structure where you have test suites and inside there are all individual test cases.
> Please add the extra layer inside test suites with test cases.

Added the extra layer with platform as the tag name.

yerabolu commented:
> > as test for the output, you should be able to write a completely independent script that loads and parses the json file and then generate the same junit files generated right now by sanitycheck.
> 
> Let me expand on this commenting on what is missing IMO.
> 
> * As a tester (and maintainer of CIs) I would like to have easily available (i.e. saved as a key-value pairs) info which is standard to test reports summarizing the results. This is an example of an entry from junit:
> 
> ```
> <testsuite errors="262" failures="0" name="nrf52840dk_nrf52840" skipped="0" tests="750" time="244.814200">
> ```
> 
> I would like to have dictionary entries for this summary info. Currently, you have time, platform, but the information with the number of failures/errors/skipped/total is missing
> 
> This should be available on all levels: e.g one can have a "master" suite, which consists of 100 test suites ("instances" in sanitycheck's code nomenclature). There should be a summary with all the above information e.g `{"errors":1, "failures":2, "skipped":5, "tests":100}`. Then each test suite (instance) will have its own summary inside, e.g. test suite #3, which has 500 test cases, will have `{"errors":0, "failures":1, "skipped":0, "tests":500}`.

Thank You for the detailed review!! Added the summary metrics for each test suite.

yerabolu commented:
> Tests can be very verbose. I would consider including handler.log only in case of failure

Added handler_logs only in case of failures.

nashif commented:
getting this:

```
i9:zephyr((HEAD detached from upstream/pull/27247)): sanitycheck -T tests/kernel/semaphore/
Renaming output directory to /home/nashif/Work/zephyrproject/zephyr/sanity-out.4
INFO    - Zephyr version: zephyr-v2.4.0-1177-gbb50eaf613
INFO    - JOBS: 20
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 20 test configurations selected, 280 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:   18/  18  100%  skipped:    1, failed:    0
INFO    - 17 of 17 tests passed (100.00%), 0 failed, 3 skipped with 0 warnings in 26.91 seconds
INFO    - In total 391 test cases were executed on 20 out of total 299 platforms (6.69%)
INFO    - 16 tests executed on platforms, 1 tests were only built.
Traceback (most recent call last):
  File "/home/nashif/Work/zephyrproject/zephyr/scripts/sanitycheck", line 1220, in <module>
    main()
  File "/home/nashif/Work/zephyrproject/zephyr/scripts/sanitycheck", line 1207, in main
    suite.save_reports(options.report_name,
  File "/home/nashif/Work/zephyrproject/zephyr/scripts/sanity_chk/sanitylib.py", line 2645, in save_reports
    self.json_report(filename + ".json", full_report=True, append=only_failed)
  File "/home/nashif/Work/zephyrproject/zephyr/scripts/sanity_chk/sanitylib.py", line 3407, in json_report
    if full_report and not instance.build_only:
AttributeError: 'TestInstance' object has no attribute 'build_only'
i9:zephyr((HEAD detached from upstream/pull/27247)): 
```

nashif commented:
getting this:

```
i9:zephyr((HEAD detached from upstream/pull/27247)): sanitycheck -T tests/kernel/semaphore/
Renaming output directory to /home/nashif/Work/zephyrproject/zephyr/sanity-out.4
INFO    - Zephyr version: zephyr-v2.4.0-1177-gbb50eaf613
INFO    - JOBS: 20
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 20 test configurations selected, 280 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:   18/  18  100%  skipped:    1, failed:    0
INFO    - 17 of 17 tests passed (100.00%), 0 failed, 3 skipped with 0 warnings in 26.91 seconds
INFO    - In total 391 test cases were executed on 20 out of total 299 platforms (6.69%)
INFO    - 16 tests executed on platforms, 1 tests were only built.
Traceback (most recent call last):
  File "/home/nashif/Work/zephyrproject/zephyr/scripts/sanitycheck", line 1220, in <module>
    main()
  File "/home/nashif/Work/zephyrproject/zephyr/scripts/sanitycheck", line 1207, in main
    suite.save_reports(options.report_name,
  File "/home/nashif/Work/zephyrproject/zephyr/scripts/sanity_chk/sanitylib.py", line 2645, in save_reports
    self.json_report(filename + ".json", full_report=True, append=only_failed)
  File "/home/nashif/Work/zephyrproject/zephyr/scripts/sanity_chk/sanitylib.py", line 3407, in json_report
    if full_report and not instance.build_only:
AttributeError: 'TestInstance' object has no attribute 'build_only'
```

nashif commented:
> @nashif We should consider having an additional "target" json report as with platform.xml for ease of publishing results. But this can be added later on and in principle should be a one-liner change.

IMO this should not be needed, the goal of the json output is not to be directly consumed or to replace current XML output, but to serve as a raw output of the sanitycheck run that can be parsed to generate anything, especially custom reports. We should not try to replicate the junit structure in some other format, I see no value here and just duplication. 

Once we have the json output that has all data, we can write custom scripts that parse the json and create reports and calculate counts and do statistics.

nashif commented:
LGTM as a baseline, we can add more data to the output as needed later.

andyross reviewed (approved):
Looks great, though I'm not the right source for source-level python review.

I guess one metawarning is that someone is likely to come in and demand a formal spec, or documentation section, or schema for exactly the format that is being generated.  Right now the only way to use this report would be ad-hoc (i.e. just read it and write your code against the live example).  Mind you, I don't **personally** think this is such a big deal.  But I strongly suspect someone does.

jenmwms commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3367,6 +3366,98 @@ def csv_report(self, filename):
+                build_log = os.path.join(instance.build_dir, "build.log")
+                device_log = os.path.join(instance.build_dir, "device.log")
+
+                handler_time = instance.metrics.get('handler_time', 0)
```
Please make this double quotes to be consistent with the other `instance.metrics.get( )`

nashif commented:
@PerMac  please re-review
